Improve internationalization

v-- opened the pull request:
While experimenting with internationalization of a personal project, I encountered two issues:
* I had to translate a bulk of messages coming from click.
* I had to configure gettext globally (or monkey-patch it) before loading click, which introduces many complications. This is discussed by another user, [carmenbianca](https://github.com/carmenbianca), in #2706.

I was initially thinking of creating a discussion about this, but I wanted to also share an implementation. I am myself inexperienced in translating software, so any criticism is welcome. I will list what I have done and what I am unsure about. Perhaps [carmenbianca](https://github.com/carmenbianca) can chime in?

* I have added some .po files with translations into Russian and Bulgarian. These are straightforward, but there are subtleties:
    * They must be compiled to .mo files. This is an issue because the setup becomes more difficult, especially on Windows where gettext tools must be installed manually.  I found no easy way to compile the .po files to .mo files. I have thus committed the .mo files, as well as a compilation script. gettext really is clunky.
    * The locale files are under the `click` domain. Client applications will have to use other gettext domains. I am unaware of best practices here.
    * In order for flit to package these files, they must be put in the source code directory.

* I have replaced the calls to gettext with a wrapper that stores a locale internally. This should not be a problem generally, but I want to hear if you have any precautions.

* I have added some unit tests that unfortunately depend on gettext. Hence, on gettext itself and the .mo files discussed above. Conditionally skipping tests in pytest is fortunately easy enough, if running them in CI (or on Windows) becomes a problem.

v-- committed:
Add localization

* Use a custom module that wraps gettext and allows setting the locale.
* Add a directory with .po files, and include the corresponding .mo files
  to avoid requiring everybody to compile them.
* Add Russian and Bulgarian localization.

davidism commented:
Thank you for working on this. I will need to very carefully review this at some point, since it affects quite a bit of the code and introduces a new pattern that needs to be maintainable later. Additionally, based on experiments for docs translations, I want to manage translations using Weblate, so I'll have to figure out how to set that up for this project. This is unlikely to happen soon, so I just want to set expectations.

v-- commented:
@davidism I appreciate your comment. I also just realized that I haven't thanked the team for developing click - it's one of the libraries I have been using consistently over the years.

Now, more to the point. I have no expectations for this pull request myself. As mentioned, the translations can easily be extracted into another package (or implemented differently in a later pull request). This would leave mostly mundane changes that should not induce maintenance overhead (replacing which implementations of `gettext` and `ngettext` get used).

Rowlando13 commented:
@v--  I appreciate your candor. I don't have a super long history with Click so sometimes it is a bit hard to tell where the line for should be with regard to accepting something. Sounds @davidism is going to help you out.

kdeldycke commented:
Browsing i18n-related PR, just realized this one has some unit tests that I was suggesting implementing in https://github.com/pallets/click/pull/2902 (by @ic).

What I see here that is left to do is:
- rebase that PR on top of `main` branch
- drop the translation files as @v-- proposed (`the translations can easily be extracted into another package (or implemented differently in a later pull request)`), which will reduce the diff, and do not interfere with @davidism future plans to rely on Weblate
- maybe adopt @carmenbianca's proposal in #2706 to implement `install_translations` to decouple the translations, but keep the current `gettext` calls as fallback so nothing changes for previous users

@v-- are you interested in rebooting the work on this PR? I can take over and try to merge your code from this PR with the good ideas from https://github.com/pallets/click/issues/2706 .

v-- commented:
@kdeldycke I have rebased my original pull request, but did not make any other changes. You can add/remove/change anything as you see fit.

One thing should be noted. The tests are failing, but the "fix" for this is not trivial. Unlike a year ago when I created the pull request, some messages such as "Show this message and exit." are now fixed and so dynamically changing the language is no longer possible without some refactoring. It's an unpleasant edge case where some strings are translated according to the new locale, but others are not.

ic commented:
@v-- Thank you for your work. Discovering your OP now, we had the exact same problems for the languages we want to support, and concur on your analysis and approach. In fact I implemented about the same code last year too! But backed off and focused on a narrower goal---first localise as many strings as possible, ideally all (#2902).

This was merged last week. I have checked on `main` for the string you take as example, and it is localised now:

```
src/click/decorators.py
548:    kwargs.setdefault("help", _("Show this message and exit."))
```

Any issue with this? (I for one forget to sync my fork with the original repo)

The PR may be missing recent new strings. It was decided to merge it for now, and likely roll forward as needed.

v-- commented:
@ic I just rebased my changes again because of a merge conflict.

Anyhow, the problem I was describing concerns the _dynamic_ change of languages. See the failing unit test.

The gist is the following line from `src/click/decorators.py`:

```python
kwargs.setdefault("help", _("Show this message and exit."))
```

The default is set _after_ getting translated, but you can nevertheless change the language following that. Then that particular translation will stick to the original language, but most strings will use the new language.

There are simple ways to mitigate this (e.g. using a `default_factory` instead of `setdefault`), but since I am not a maintainer, I have no preference for whether to support dynamic translation at all.

ic commented:
Thanks @v-- , now very clear to me what the problem is. It seems the issues are bound to `src/click/decorators.py` (one way to deal with it could be to turn them into callables).

I understand this breaks tests _in the same runtime_, but I wonder whether it is a problem in practice. It could when the CLI is interactive and a user somehow changes the locale in the shell executing the CLI, but is it likely?

CLI are often (mostly?) one shot runs, where the execution environment does not change (much). So wondering whether there is a problem, even though the breaking test is technically correct.
